sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java

```diff
@@ -839,43 +902,70 @@ private void newTimeout(
 
-        if (future.isSuccess()) {
+        try {
+            if (future.isSuccess()) {
+                final Channel channel = future.channel();
 
-            final Channel channel = future.channel();
+                try {
+                    this.poolHandler.channelAcquired(channel);
+                } catch (Throwable error) {
+                    this.closeChannelAndFail(channel, error, promise);
+                    return;
+                }
 
-            try {
-                this.poolHandler.channelAcquired(channel);
-            } catch (Throwable error) {
-                this.closeChannelAndFail(channel, error, promise);
-                return;
-            }
+                if (promise.trySuccess(channel)) {
+
+                    if (logger.isDebugEnabled()) {
+                        logger.debug("established a channel local {}, remote {}", channel.localAddress(), channel.remoteAddress());
+                    }
+                    this.acquiredChannels.compute(channel, (ignored, acquiredChannel) -> {
+                        //                    reportIssueUnless(logger, v == null, this, "expected null channel, not {}", v);
+
+                        reportIssueUnless(logger, acquiredChannel == null, this,
+                            "Channel({}) to be acquired has already been acquired",
+                            channel);
+                        reportIssueUnless(logger, !this.availableChannels.remove(channel), this,
+                            "Channel({}) to be acquired is still in the list of available channels",
+                            channel);
+
+                        return channel;
+                    });
+                } else {
+                    if (logger.isDebugEnabled()) {
+                        logger.debug("notifyChannelConnect promise.trySuccess(channel)=false");
+                    }
+                    // Promise was completed in the meantime (like cancelled), just close the channel
+                    this.closeChannel(channel);
+                }
 
-            if (promise.trySuccess(channel)) {
-                this.acquiredChannels.compute(channel, (k, v) -> {
-                    reportIssueUnless(logger, v == null, this, "expected null channel, not {}", v);
-                    this.connecting.set(false);
-                    return channel;
-                });
             } else {
-                // Promise was completed in the meantime (like cancelled), just close the channel
-                this.closeChannel(channel);
-                this.connecting.set(false);
+                if (logger.isDebugEnabled()) {
+                    logger.debug("notifyChannelConnect future was not successful");
+                }
+                promise.tryFailure(future.cause());
             }
-
-        } else {
-            promise.tryFailure(future.cause());
+        } finally {
+            this.connecting.set(false);
```

Comment:
+1 to this pattern. 

earlier might look okey as poolHandler notification is very light weight and almost doesn't fail, but this is a pattern we should force in reviews. 
Thanks,